[Feat] 알림함에서 초대 수락 시 네비게이션 플로우 및 초대 수락/거절 관련 통신 코드를 구현했습니다.

Rookie0031 opened the pull request:
## 관련 이슈
- closed #97 

## 배경
 유저가 알림함에서 초대를 수락할 경우에, 포지션을 선택하고 밴드가입이 완료되게 됩니다. 아래와 같은 디자인을 구현했습니다.
<img width="300" alt="image" src="https://user-images.githubusercontent.com/103009135/222439630-06f98e90-52e3-468c-992f-85029fdaccb7.png">

## 스크린샷
https://user-images.githubusercontent.com/103009135/222450213-9e5f53e7-febc-48c8-830b-e4f0c1f8ca6b.mov

## 테스트 방법
진입점을 아래처럼 바꿔주세요!

        window.rootViewController = UINavigationController(rootViewController: NotificationListViewController())


## 작업 내용

### 1️⃣ BandCreationFinishGuideViewController 
#### 가입 완료 안내 VC 입니다. 
안내 문구 내에 이미지가 들어가기 때문에 @ccdkk 가 작성해주신 안내용 모달 기본틀 내에서 적절히 변형하여 사용했습니다.
스택내에 스택이 있는 구조로 그림과 같습니다.
<img width="600" alt="image" src="https://user-images.githubusercontent.com/103009135/222442291-88a40c17-c180-445e-8be8-7dc69aceac8b.png">
</br>

### 2️⃣ PositionSelectForInvitationViewController
 초대 수락 후, 밴드에서 자신의 포지션을 선택하는 VC 입니다. 
공통 컴포넌트 사항이지만 분기처리가 번거로워 따로 파일을 만들었습니다.
추후에 유저의 정보를 조회해서 유저가 선택한 가능한 악기들이 선택 옵션으로 업데이트될 예정입니다.
</br>


### 3️⃣ 초대 수락시 PUT
 초대를 수락할 경우, 필요한 쿼리 컴포넌트와 함께 PUT 요청을 합니다. @ccdkk 가 구현해주신 PUT 함수와 거의 유사합니다.

https://github.com/extreme-rock/GetARock-iOS/blob/904f542f22f3786d25a127c1216695e921be9763/GetARock/GetARock/Network/NotificationNetworkManager.swift#L42-L84

#### 초대 수락시 쿼리 컴포넌트에는 악기 리스트가 보내지며 각 element는 , 로 구분됩니다.
따라서 이 형태에 맞게 문자열을 만들기 위한 함수가 추가되었습니다.
https://github.com/extreme-rock/GetARock-iOS/blob/904f542f22f3786d25a127c1216695e921be9763/GetARock/GetARock/Screens/NotificationList/PositionSelectForInvitationViewController.swift#L81-L85
</br>

### 4️⃣ 초대 거절시 PUT 
 초대를 거절한 경우에도 유사한 형태의 통신 로직이 쓰였습니다.
https://github.com/extreme-rock/GetARock-iOS/blob/904f542f22f3786d25a127c1216695e921be9763/GetARock/GetARock/Network/NotificationNetworkManager.swift#L86-L126
</br>

## 리뷰 노트
- 불필요한 코드
- 코드 컨벤션 미준수 사항
- 디자인 미준수 사항
- 개선 가능한 로직

<!-- 해당 PR 리뷰 과정에서 논의된 확정 사항을 develop 브랜치에 머지하기 전 정리합니다. -->

Rookie0031 committed:
[Feat] 초대 수락시 PUT 요청 함수 추가 (#97)

Rookie0031 committed:
[Feat] 초대 거절 시 PUT 요청 함수 추가 (#97)

Rookie0031 committed:
[Add] 통신 함수 VC내에 추가 및 주석 작업 (#97)

Rookie0031 committed:
[Design] 초대 수락시 포지션 선택 VC 구현 (#97)

Rookie0031 committed:
[Add] 이미지 리터럴 및 이미지 파일 추가 (#97)

Rookie0031 committed:
[Design] 밴드 생성 완료 후 환영 페이지 구현 (#97)

Rookie0031 committed:
[Design] BasicLabel에 두 글자 이상일 경우 행간 추가 (#97)

Rookie0031 committed:
[Feat] 가입환영 VC 연결 (#97)

Rookie0031 committed:
[Comment] 주석 작업 추가 (#97)

Rookie0031 committed:
[Feat] Navigation에서 Modal 로 전환 (#97)

Rookie0031 commented on GetARock/GetARock/Network/NotificationNetworkManager.swift:
```diff
@@ -38,8 +39,89 @@ final class NotificationNetworkManager {
+        let alertIdQuery = URLQueryItem(name: "alertId", value: String(alertId))
+        let bandIdQuery = URLQueryItem(name: "bandId", value: String(bandId))
+        let memberIdQuery = URLQueryItem(name: "memberId", value: String(memberId))
+        let instrumentQuery = URLQueryItem(name: "memberInstrument", value: memberInstrument.joined(separator: ","))
```
선택 악기가 여러개인 경우, 파라미터로 문자열 배열을 받지만 쿼리는 하나의 문자열로 보내져야합니다.
그래서 , 로 구분되게 만들기 위해서 joined 메소드를 사용합니다

Rookie0031 commented on GetARock/GetARock/Network/NotificationNetworkManager.swift:
```diff
@@ -38,8 +39,89 @@ final class NotificationNetworkManager {
+        
+        queryURLComponent?.queryItems = [alertIdQuery, bandIdQuery, memberIdQuery, instrumentQuery]
+        guard let url = queryURLComponent?.url else {
+            print(NetworkError.badURL.errorDescription ?? "")
```
이 함수는 UIAction의 클로저 파라미터에 들어가는데, 그 클로저의 형태가 throw 함수는 받지 못하게 되어있습니다.
그래서 이 부분은 에러 throw가 아니라 print로 대체했습니다.

Rookie0031 commented on GetARock/GetARock/Network/NotificationNetworkManager.swift:
```diff
@@ -7,6 +7,7 @@
+//TODO: 추후 프린트문 error handling case 체크 필요
 final class NotificationNetworkManager {
     
     static let shared = NotificationNetworkManager()
```
Singleton 패턴의 단일인스턴스로 함수를 만드는 경우에 이 함수들이 사용되는 VC가 deinit 되면 인스턴스도 함께 메모리에서 내려가기 때문에 메모리 누수를 방지할 수 있다는 장점이 있다고 하네요,,
static func으로 선언할 경우, 프로세스 작동시간 동안 계속 메모리를 잡아먹기 때문에.. 네트워크 함수는 인스턴스 메소드로 만드는게 좀 더 적절한거같아요!!

ccdkk commented on GetARock/GetARock/Global/Utils/BandCreationFinishGuideViewController.swift:
```diff
@@ -0,0 +1,195 @@
+                 fontStyle: .content,
+                 textColorInfo: .white))
+    
+    private lazy var labelStackView: UIStackView = {
```
p4;
titleStackView는 어떠신가요? 
아래에도 Label들이 많아서 어떤 라벨을 포함하는 스택뷰인지 덜 직관적일 수 있을 것 같아요

Rookie0031 replied:
그러네요 감사합니다 반영완료!  https://github.com/extreme-rock/GetARock-iOS/pull/149/commits/8d5df397d554ab275405049212677b8193b6c6e3

ccdkk commented on GetARock/GetARock/Global/Utils/BandCreationFinishGuideViewController.swift:
```diff
@@ -0,0 +1,195 @@
+        super.viewDidLoad()
+        self.attribute()
+        self.setupLayout()
+        self.navigationController?.navigationBar.isHidden = true
```
p3;
모달로 띄워지는 부분이라 없어도 되지 않을까용?

Rookie0031 replied:
헉 그러네요 감사합니다! 반영완료!  https://github.com/extreme-rock/GetARock-iOS/pull/149/commits/8d5df397d554ab275405049212677b8193b6c6e3

ccdkk commented on GetARock/GetARock/Network/NotificationNetworkManager.swift:
```diff
@@ -38,8 +39,89 @@ final class NotificationNetworkManager {
+        request.httpMethod = "PUT"
+        request.allHTTPHeaderFields = headers
+        
+        let dataTask = URLSession.shared.dataTask(with: request as URLRequest) { data, response, error in
```
p2;
URLRequest로의 타입캐스팅이 없어도 될 것 같아용

Rookie0031 replied:
오 감사합니다! 반영완료!  https://github.com/extreme-rock/GetARock-iOS/pull/149/commits/a88b295f90ee398b22e939dbe5d09e2d025ee43d

ccdkk commented on GetARock/GetARock/Global/Utils/BandCreationFinishGuideViewController.swift:
```diff
@@ -0,0 +1,195 @@
+                                     bandTimelineHstack,
+                                     commentHstack]))
+    
+    private let confirmButton: BottomButton = {
+        $0.setTitle("확인 완료 😎", for: .normal)
+        $0.titleLabel?.font = .setFont(.headline02)
+        $0.titleLabel?.textColor = .white
+        return $0
+    }(BottomButton())
```
q;
확인 완료 버튼을 누르면 들어가는 인터랙션은 다음 PR에서 구현되나용?

Rookie0031 replied:
아 이 부분은 밴드 상세 정보 뷰 구조가 완성되면 작업하는게 나을 거 같아서 안했습니다.
완료하면 밴드 상세 정보 뷰로 넘어가는데, 그 뷰에 전달해줄 데이터 (ex. 밴드 id)가 생기지 않을까해서 나중에 작업하는게 좀 더 나을 거 같아서요. 그냥 알아보기 쉽게 연결해놓는거도 좋을까요?

ccdkk replied:
TODO로 남겨놓고 다음에 작업해도 괜찮을 것 같아요!

ccdkk commented on GetARock/GetARock/Screens/NotificationList/PositionSelectForInvitationViewController.swift:
```diff
@@ -0,0 +1,108 @@
+        let action = UIAction { _ in self.navigateToNext() }
+        $0.addAction(action, for: .touchUpInside)
+        return $0
+    }(BottomButton())
```
p3;
let으로 구현해도될 것 같아요!

Rookie0031 replied:
position을 말씀하신거죠?? 반영완료!!  https://github.com/extreme-rock/GetARock-iOS/pull/149/commits/8d5df397d554ab275405049212677b8193b6c6e3

ccdkk replied:
의도는 여기 UI들을 말한거였는데 lazy군요 ㅋㅋ 잘못봤습니당

ccdkk reviewed (commented):
고생하셨습니다!!

ccdkk commented on GetARock/GetARock/Screens/NotificationList/PositionSelectForInvitationViewController.swift:
```diff
@@ -0,0 +1,108 @@
+        let nextVC = BandCreationFinishGuideViewController()
+        present(nextVC, animated: true)
+    }
+}
```
p3;
acceptInvitation이 성공할 경우에 nextVC를 띄워주는게 맞을 것 같은데 이 부분을 escaping closure로 구현하는 것은 어떤가요?
escaping closure에 파라미터로 성공 여부(?) 를 전달하고 이에 따른 분기처리가 필요해 보입니다.

Rookie0031 replied:
오오오 상당히 좋은 포인트네요...갓갓 감사합니다!! 👍 반영완료!! https://github.com/extreme-rock/GetARock-iOS/pull/149/commits/1bb9651cec1d176dee673a6613889a6bdca3fbac

ccdkk replied:
실패할 경우에는 어떻게 해야할지 TODO로 남겨놓으면 좋을 것 같아요 !

Rookie0031 committed:
[Style] 컨벤션 미준수 사항 수정 (#97)

Rookie0031 committed:
[Style] 불필요한 타입 캐스팅 삭제 (#97)

Rookie0031 committed:
[Style] escaping clousre 삽입을 통한 코드 가독성 개선 (#97)

GODNOEL commented on GetARock/GetARock/Network/NotificationNetworkManager.swift:
```diff
@@ -38,8 +39,90 @@ final class NotificationNetworkManager {
+                case (200...299):
+                    print("success")
+                    print(httpResponse)
+                    DispatchQueue.main.async { completion() } // completion 클로저가 UI 관련 코드를 포함함
```
p
acacceptInvitation에서 파라미터로 completion()를 받고 있던데 얘는 뭐하는 친구인가요?
데이터를 put하고 나서 이후에 프론트에서 해야할 일을 담은 함수인가요?

Rookie0031 replied:
네 맞아요 completion은 클로저인데 response code가 200번대 즉, POST가 성공할 경우에만 수행됩니다.

        NotificationNetworkManager.shared.acceptInvitation(
            alertId: 0,
            bandId: 0,
            memberId: 0,
            memberInstrument: getSelectedInstrumentList(),
            completion: {
                self.present(BandCreationFinishGuideViewController(), animated: true)
            })

이런식으로 함수를 사용하는 단계에서 POST가 성공하나 경우에 어떤 코드를 수행할지 넘겨줄 수 있어요! 
❗️ 특이한점으로, 저 부분에서 사용된 클로저는 escaping closure입니다.

클로저가 함수나 메소드의 인자로 전달되어 사용될 때, 그 클로저가 함수나 메소드가 반환(return)되는 시점보다 더 오래 살아남아야 할 경우가 있습니다. 이를 "escaping closure"라고 부릅니다. 예를 들어, 비동기 작업이 완료될 때 호출되는 클로저는 함수가 반환된 이후에도 실행될 수 있기 때문에 escaping closure로 분류됩니다.

dataTask 함수는 아직까지 async await를 지원하지 않는거로 알고 있어요. 그래서 수행이 끝날때까지 기다리라고 컴파일 시점에서 명시할 수 없기에 escaping closure로 함수가 끝나도 수행되게끔 처리한것입니다

GODNOEL commented on GetARock/GetARock/Screens/NotificationList/PositionSelectForInvitationViewController.swift:
```diff
@@ -0,0 +1,108 @@
+        let action = UIAction { _ in self.navigateToNext() }
+        $0.addAction(action, for: .touchUpInside)
+        return $0
+    }(BottomButton())
```
p:
요기 action에는 `[weak self] _ in` 처리를 안하신 이유가 있나요?

P3;
지금은 포지션을 선택하지않고 가입하기를 선택해도 버튼이 작동하는데
처음 선택된 Cell이 없으면 버튼이 disable  상태여야할것같아요..!

GODNOEL commented on GetARock/GetARock/Screens/NotificationList/PositionSelectForInvitationViewController.swift:
```diff
@@ -0,0 +1,108 @@
+        for index in startIndex..<endIndex {
+            positionCollectionView.updateCellIndex(at: IndexPath(item: index, section: 0))
+        }
+    }
```
p;
요 친구는 왜 cell의 인덱스를 순회하면서 업데이트하나요..?

Rookie0031 replied:
이건 이전에 데이크가 만든 코드를 복붙하면서 불필요한 부분을 가져왔네요 감사합니다!!!  https://github.com/extreme-rock/GetARock-iOS/pull/149/commits/c90796b2f33f0597a9de59133510875a1e01575b

GODNOEL reviewed (commented):
### 고생하셧습니다 루키!
어려운 플로우인데 너무 잘 작업해주셔서 감사하네요! 👍 갓루키

근데 모여락이 빠지면서 `BandCreationFinishGuideViewController`는 추후에 사용될 것 같아요..! 
지금 디자인에서는 밴드를 만들면 모달이 아닌 팝업이 뜨는데 해당 플로우 변경된거 한스가 안내해주셨나요..?

그리고 PR과는 관계없지만 밴드 요청을 거절하면 뜨는 팝업에서 선택지가 둘다 거절인데
좌측의 `확인` -> `취소`로 변경해주시면 좋을것같아요..!

<img width="280" alt="image" src="https://user-images.githubusercontent.com/103012104/223614479-662dc654-0b6c-4156-9094-88bcd389def2.png">
<img width="600" alt="image" src="https://user-images.githubusercontent.com/103012104/223614370-58cd9b09-20a4-4116-bcd7-757b931d1814.png">

<br>

글고 이것도 이번 PR하고 상관 없는건데..
디자인에 맞춰서 거절 버튼 UI도 수락에 비해 눈에 안띄게 수정해주실 수 있나요?
저번 PR때 잘 안보고 늦게 말해서 죄송합니다ㅠㅠ
거절 버튼 배경 : dark02 / 외곽선 : gray2
<img width="300" alt="image" src="https://user-images.githubusercontent.com/103012104/223615494-11399b91-e4b2-4056-95f6-2b25855af7e2.png">

ccdkk reviewed (approved):
고생하셨습니다!!!

Rookie0031 committed:
[Comment] 주석 작업 추가 (#97)

Rookie0031 committed:
[Design] 거절 버튼 색상 변경 및 Alert 문구 변경 (#97)

Rookie0031 committed:
[Style] weak self 구문 추가 (#97)

Rookie0031 committed:
[Del] 불필요한 코드 삭제 (#97)

Rookie0031 commented:
> ### 고생하셧습니다 루키!
> 어려운 플로우인데 너무 잘 작업해주셔서 감사하네요! 👍 갓루키
> 
> 근데 모여락이 빠지면서 `BandCreationFinishGuideViewController`는 추후에 사용될 것 같아요..! 지금 디자인에서는 밴드를 만들면 모달이 아닌 팝업이 뜨는데 해당 플로우 변경된거 한스가 안내해주셨나요..?
> 
> 그리고 PR과는 관계없지만 밴드 요청을 거절하면 뜨는 팝업에서 선택지가 둘다 거절인데 좌측의 `확인` -> `취소`로 변경해주시면 좋을것같아요..!
> 
> <img alt="image" width="280" src="https://user-images.githubusercontent.com/103012104/223614479-662dc654-0b6c-4156-9094-88bcd389def2.png">
> 
> <img alt="image" width="600" src="https://user-images.githubusercontent.com/103012104/223614370-58cd9b09-20a4-4116-bcd7-757b931d1814.png">
> 
> 
> 글고 이것도 이번 PR하고 상관 없는건데.. 디자인에 맞춰서 거절 버튼 UI도 수락에 비해 눈에 안띄게 수정해주실 수 있나요? 저번 PR때 잘 안보고 늦게 말해서 죄송합니다ㅠㅠ 거절 버튼 배경 : dark02 / 외곽선 : gray2 <img alt="image" width="300" src="https://user-images.githubusercontent.com/103012104/223615494-11399b91-e4b2-4056-95f6-2b25855af7e2.png">

너무 좋은데요! 바로 반영했어요 감사합니다! https://github.com/extreme-rock/GetARock-iOS/pull/149/commits/dc4a61fa91d34d05366656f790c9dc33b21a4e16

Rookie0031 committed:
[Feat] Position 선택에 따라 버튼 활성, 비활성 기능 추가 (#97)

1. 공통 컴포넌트인 PositionCollectionView에 버튼 select, deselect에 따른 Notification Post 구조 추가
2. 초대 수락시 포지션 선택 VC에 Notification Observer 추가
3. Notification 수신에 따른 버튼 활성, 비활성 함수 추가

Rookie0031 commented:
> > ### 고생하셧습니다 루키!
> > 어려운 플로우인데 너무 잘 작업해주셔서 감사하네요! 👍 갓루키
> > 근데 모여락이 빠지면서 `BandCreationFinishGuideViewController`는 추후에 사용될 것 같아요..! 지금 디자인에서는 밴드를 만들면 모달이 아닌 팝업이 뜨는데 해당 플로우 변경된거 한스가 안내해주셨나요..?
> > 그리고 PR과는 관계없지만 밴드 요청을 거절하면 뜨는 팝업에서 선택지가 둘다 거절인데 좌측의 `확인` -> `취소`로 변경해주시면 좋을것같아요..!
> > <img alt="image" width="280" src="https://user-images.githubusercontent.com/103012104/223614479-662dc654-0b6c-4156-9094-88bcd389def2.png">
> > <img alt="image" width="600" src="https://user-images.githubusercontent.com/103012104/223614370-58cd9b09-20a4-4116-bcd7-757b931d1814.png">
> > 글고 이것도 이번 PR하고 상관 없는건데.. 디자인에 맞춰서 거절 버튼 UI도 수락에 비해 눈에 안띄게 수정해주실 수 있나요? 저번 PR때 잘 안보고 늦게 말해서 죄송합니다ㅠㅠ 거절 버튼 배경 : dark02 / 외곽선 : gray2 <img alt="image" width="300" src="https://user-images.githubusercontent.com/103012104/223615494-11399b91-e4b2-4056-95f6-2b25855af7e2.png">
> 
> 너무 좋은데요! 바로 반영했어요 감사합니다! [dc4a61f](https://github.com/extreme-rock/GetARock-iOS/commit/dc4a61fa91d34d05366656f790c9dc33b21a4e16)

추가로 악기 선택에 따른 버튼 활성, 비활성 기능도 추가 했어요! 
https://github.com/extreme-rock/GetARock-iOS/pull/149/commits/918a88f6cae6e8a88aa497623a16053b10eb13f6

Rookie0031 committed:
Merge branch 'develop' into feature/97-invitation-accept-and-reject

Rookie0031 committed:
[Style] merge 과정에서 발생한 conflict 해결 (#97)

GODNOEL reviewed (approved):
고생하셨어요!! 룩희!! 
어프룹 도장 꽝‼️ 👍

Rookie0031 committed:
[Design] 모여락 기능 삭제에 따른 안내창 모달 띄우기 UI 삭제 (#97)
